File: libsignal-service/src/main/java/org/whispersystems/signalservice/api/crypto/AttachmentCipherInputStream.java
/*
 * Copyright (C) 2014-2017 Open Whisper Systems
 *
 * Licensed according to the LICENSE file in this repository.
 */

package org.whispersystems.signalservice.api.crypto;

import org.signal.core.util.stream.LimitedInputStream;
import org.signal.libsignal.protocol.InvalidMessageException;
import org.signal.libsignal.protocol.incrementalmac.ChunkSizeChoice;
import org.signal.libsignal.protocol.incrementalmac.IncrementalMacInputStream;
import org.signal.libsignal.protocol.kdf.HKDF;
import org.whispersystems.signalservice.api.backup.BackupKey;
import org.whispersystems.signalservice.internal.util.Util;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FilterInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.Mac;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.ShortBufferException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * Class for streaming an encrypted push attachment off disk.
 *
 * @author Moxie Marlinspike
 */

public class AttachmentCipherInputStream extends FilterInputStream {

  private static final int BLOCK_SIZE      = 16;
  private static final int CIPHER_KEY_SIZE = 32;
  private static final int MAC_KEY_SIZE    = 32;

  private final Cipher cipher;
  private final long   totalDataSize;

  private boolean done;
  private long    totalRead;
  private byte[]  overflowBuffer;

  /**
   * Passing in a null incrementalDigest and/or 0 for the chunk size at the call site disables incremental mac validation.
   */
  public static LimitedInputStream createForAttachment(File file, long plaintextLength, byte[] combinedKeyMaterial, byte[] digest, byte[] incrementalDigest, int incrementalMacChunkSize)
      throws InvalidMessageException, IOException {
    return createForAttachment(file, plaintextLength, combinedKeyMaterial, digest, incrementalDigest, incrementalMacChunkSize, false);
  }

  /**
   * Passing in a null incrementalDigest and/or 0 for the chunk size at the call site disables incremental mac validation.
   *
   * Passing in true for ignoreDigest DOES NOT VERIFY THE DIGEST
   */
  public static LimitedInputStream createForAttachment(File file, long plaintextLength, byte[] combinedKeyMaterial, byte[] digest, byte[] incrementalDigest, int incrementalMacChunkSize, boolean ignoreDigest)
      throws InvalidMessageException, IOException
  {
    return createForAttachment(() -> new FileInputStream(file), file.length(), plaintextLength, combinedKeyMaterial, digest, incrementalDigest, incrementalMacChunkSize, ignoreDigest);
  }

  /**
   * Passing in a null incrementalDigest and/or 0 for the chunk size at the call site disables incremental mac validation.
   *
   * Passing in true for ignoreDigest DOES NOT VERIFY THE DIGEST
   */
  public static LimitedInputStream createForAttachment(StreamSupplier streamSupplier, long streamLength, long plaintextLength, byte[] combinedKeyMaterial, byte[] digest, byte[] incrementalDigest, int incrementalMacChunkSize, boolean ignoreDigest)
      throws InvalidMessageException, IOException
  {
    byte[][] parts = Util.split(combinedKeyMaterial, CIPHER_KEY_SIZE, MAC_KEY_SIZE);
    Mac      mac   = initMac(parts[1]);

    if (streamLength <= BLOCK_SIZE + mac.getMacLength()) {
      throw new InvalidMessageException("Message shorter than crypto overhead! length: " + streamLength);
    }

    if (!ignoreDigest && digest == null) {
      throw new InvalidMessageException("Missing digest!");
    }

    final InputStream wrappedStream;
    final boolean     hasIncrementalMac = incrementalDigest != null && incrementalDigest.length > 0 && incrementalMacChunkSize > 0;

    if (!hasIncrementalMac) {
      try (InputStream macVerificationStream = streamSupplier.openStream()) {
        verifyMac(macVerificationStream, streamLength, mac, digest);
      }
      wrappedStream = streamSupplier.openStream();
    } else {
      wrappedStream = new IncrementalMacInputStream(
          new IncrementalMacAdditionalValidationsInputStream(
              streamSupplier.openStream(),
              streamLength,
              mac,
              digest
          ),
          parts[1],
          ChunkSizeChoice.everyNthByte(incrementalMacChunkSize),
          incrementalDigest);
    }
    InputStream inputStream = new AttachmentCipherInputStream(wrappedStream, parts[0], streamLength - BLOCK_SIZE - mac.getMacLength());

    if (plaintextLength != 0) {
      return new LimitedInputStream(inputStream, plaintextLength);
    } else {
      return LimitedInputStream.withoutLimits(inputStream);
    }
  }

  /**
   * Decrypt archived media to it's original attachment encrypted blob.
   */
  public static LimitedInputStream createForArchivedMedia(BackupKey.MediaKeyMaterial archivedMediaKeyMaterial, File file, long originalCipherTextLength)
      throws InvalidMessageException, IOException
  {
    Mac mac = initMac(archivedMediaKeyMaterial.getMacKey());

    if (file.length() <= BLOCK_SIZE + mac.getMacLength()) {
      throw new InvalidMessageException("Message shorter than crypto overhead!");
    }

    try (FileInputStream macVerificationStream = new FileInputStream(file)) {
      verifyMac(macVerificationStream, file.length(), mac, null);
    }

    InputStream inputStream = new AttachmentCipherInputStream(new FileInputStream(file), archivedMediaKeyMaterial.getCipherKey(), file.length() - BLOCK_SIZE - mac.getMacLength());

    if (originalCipherTextLength != 0) {
      return new LimitedInputStream(inputStream, originalCipherTextLength);
    } else {
      return LimitedInputStream.withoutLimits(inputStream);
    }
  }

  public static LimitedInputStream createStreamingForArchivedAttachment(BackupKey.MediaKeyMaterial archivedMediaKeyMaterial, File file, long originalCipherTextLength, long plaintextLength, byte[] combinedKeyMaterial, byte[] digest, byte[] incrementalDigest, int incrementalMacChunkSize)
      throws InvalidMessageException, IOException
  {
    final InputStream archiveStream = createForArchivedMedia(archivedMediaKeyMaterial, file, originalCipherTextLength);

    byte[][] parts = Util.split(combinedKeyMaterial, CIPHER_KEY_SIZE, MAC_KEY_SIZE);
    Mac      mac   = initMac(parts[1]);

    if (originalCipherTextLength <= BLOCK_SIZE + mac.getMacLength()) {
      throw new InvalidMessageException("Message shorter than crypto overhead!");
    }

    if (digest == null) {
      throw new InvalidMessageException("Missing digest!");
    }

    final InputStream wrappedStream;
      wrappedStream = new IncrementalMacInputStream(
          new IncrementalMacAdditionalValidationsInputStream(
              archiveStream,
              file.length(),
              mac,
              digest
          ),
          parts[1],
          ChunkSizeChoice.everyNthByte(incrementalMacChunkSize),
          incrementalDigest);

    InputStream inputStream = new AttachmentCipherInputStream(wrappedStream, parts[0], file.length() - BLOCK_SIZE - mac.getMacLength());

    if (plaintextLength != 0) {
      return new LimitedInputStream(inputStream, plaintextLength);
    } else {
      return LimitedInputStream.withoutLimits(inputStream);
    }

  }

  public static InputStream createForStickerData(byte[] data, byte[] packKey)
      throws InvalidMessageException, IOException
  {
    byte[]   combinedKeyMaterial = HKDF.deriveSecrets(packKey, "Sticker Pack".getBytes(), 64);
    byte[][] parts               = Util.split(combinedKeyMaterial, CIPHER_KEY_SIZE, MAC_KEY_SIZE);
    Mac      mac                 = initMac(parts[1]);

    if (data.length <= BLOCK_SIZE + mac.getMacLength()) {
      throw new InvalidMessageException("Message shorter than crypto overhead!");
    }

    try (InputStream inputStream = new ByteArrayInputStream(data)) {
      verifyMac(inputStream, data.length, mac, null);
    }

    return new AttachmentCipherInputStream(new ByteArrayInputStream(data), parts[0], data.length - BLOCK_SIZE - mac.getMacLength());
  }

  private AttachmentCipherInputStream(InputStream inputStream, byte[] cipherKey, long totalDataSize)
      throws IOException
  {
    super(inputStream);

    try {
      byte[] iv = new byte[BLOCK_SIZE];
      readFully(iv);

      this.cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
      this.cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(cipherKey, "AES"), new IvParameterSpec(iv));

      this.done          = false;
      this.totalRead     = 0;
      this.totalDataSize = totalDataSize;
    } catch (NoSuchAlgorithmException | InvalidKeyException | NoSuchPaddingException | InvalidAlgorithmParameterException e) {
      throw new AssertionError(e);
    }
  }

  @Override
  public int read() throws IOException {
    byte[] buffer = new byte[1];
    int    read;

    //noinspection StatementWithEmptyBody
    while ((read = read(buffer)) == 0) ;

    return (read == -1) ? -1 : ((int) buffer[0]) & 0xFF;
  }

  @Override
  public int read(@Nonnull byte[] buffer) throws IOException {
    return read(buffer, 0, buffer.length);
  }

  @Override
  public int read(@Nonnull byte[] buffer, int offset, int length) throws IOException {
    if (totalRead != totalDataSize) {
      return readIncremental(buffer, offset, length);
    } else if (!done) {
      return readFinal(buffer, offset, length);
    } else {
      return -1;
    }
  }

  @Override
  public boolean markSupported() {
    return false;
  }

  @Override
  public long skip(long byteCount) throws IOException {
    long skipped = 0L;
    while (skipped < byteCount) {
      byte[] buf  = new byte[Math.min(4096, (int) (byteCount - skipped))];
      int    read = read(buf);

      skipped += read;
    }

    return skipped;
  }

  private int readFinal(byte[] buffer, int offset, int length) throws IOException {
    try {
      byte[] internal     = new byte[buffer.length];
      int    actualLength = Math.min(length, cipher.doFinal(internal, 0));
      System.arraycopy(internal, 0, buffer, offset, actualLength);

      done = true;
      return actualLength;
    } catch (IllegalBlockSizeException | BadPaddingException | ShortBufferException e) {
      throw new IOException(e);
    }
  }

  private int readIncremental(byte[] buffer, int offset, int length) throws IOException {
    int readLength = 0;
    if (null != overflowBuffer) {
      if (overflowBuffer.length > length) {
        System.arraycopy(overflowBuffer, 0, buffer, offset, length);
        overflowBuffer = Arrays.copyOfRange(overflowBuffer, length, overflowBuffer.length);
        return length;
      } else if (overflowBuffer.length == length) {
        System.arraycopy(overflowBuffer, 0, buffer, offset, length);
        overflowBuffer = null;
        return length;
      } else {
        System.arraycopy(overflowBuffer, 0, buffer, offset, overflowBuffer.length);
        readLength += overflowBuffer.length;
        offset += readLength;
        length -= readLength;
        overflowBuffer = null;
      }
    }

    if (length + totalRead > totalDataSize)
      length = (int) (totalDataSize - totalRead);

    byte[] internalBuffer = new byte[length];
    int    read           = super.read(internalBuffer, 0, internalBuffer.length <= cipher.getBlockSize() ? internalBuffer.length : internalBuffer.length - cipher.getBlockSize());
    totalRead += read;

    try {
      int outputLen = cipher.getOutputSize(read);

      if (outputLen <= length) {
        readLength += cipher.update(internalBuffer, 0, read, buffer, offset);
        return readLength;
      }

      byte[] transientBuffer = new byte[outputLen];
      outputLen = cipher.update(internalBuffer, 0, read, transientBuffer, 0);
      if (outputLen <= length) {
        System.arraycopy(transientBuffer, 0, buffer, offset, outputLen);
        readLength += outputLen;
      } else {
        System.arraycopy(transientBuffer, 0, buffer, offset, length);
        overflowBuffer = Arrays.copyOfRange(transientBuffer, length, outputLen);
        readLength += length;
      }
      return readLength;
    } catch (ShortBufferException e) {
      throw new AssertionError(e);
    }
  }

  private static Mac initMac(byte[] key) {
    try {
      Mac mac = Mac.getInstance("HmacSHA256");
      mac.init(new SecretKeySpec(key, "HmacSHA256"));
      return mac;
    } catch (NoSuchAlgorithmException | InvalidKeyException e) {
      throw new AssertionError(e);
    }
  }

  private static void verifyMac(@Nonnull InputStream inputStream, long length, @Nonnull Mac mac, @Nullable byte[] theirDigest)
      throws InvalidMessageException
  {
    try {
      MessageDigest digest        = MessageDigest.getInstance("SHA256");
      int           remainingData = Util.toIntExact(length) - mac.getMacLength();
      byte[]        buffer        = new byte[4096];

      while (remainingData > 0) {
        int read = inputStream.read(buffer, 0, Math.min(buffer.length, remainingData));
        mac.update(buffer, 0, read);
        digest.update(buffer, 0, read);
        remainingData -= read;
      }

      byte[] ourMac   = mac.doFinal();
      byte[] theirMac = new byte[mac.getMacLength()];
      Util.readFully(inputStream, theirMac);

      if (!MessageDigest.isEqual(ourMac, theirMac)) {
        throw new InvalidMessageException("MAC doesn't match!");
      }

      byte[] ourDigest = digest.digest(theirMac);

      if (theirDigest != null && !MessageDigest.isEqual(ourDigest, theirDigest)) {
        throw new InvalidMessageException("Digest doesn't match!");
      }

    } catch (IOException | ArithmeticException e1) {
      throw new InvalidMessageException(e1);
    } catch (NoSuchAlgorithmException e) {
      throw new AssertionError(e);
    }
  }

  private void readFully(byte[] buffer) throws IOException {
    int offset = 0;

    for (; ; ) {
      int read = super.read(buffer, offset, buffer.length - offset);

      if (read + offset < buffer.length) {
        offset += read;
      } else {
        return;
      }
    }
  }

  public interface StreamSupplier {
    @Nonnull InputStream openStream() throws IOException;
  }
}

File: libsignal-service/src/main/java/org/whispersystems/signalservice/api/crypto/AttachmentCipherInputStream.kt
/*
 * Copyright (C) 2014-2017 Open Whisper Systems
 *
 * Licensed according to the LICENSE file in this repository.
 */
package org.whispersystems.signalservice.api.crypto

import org.signal.core.util.Base64
import org.signal.core.util.readNBytesOrThrow
import org.signal.core.util.stream.LimitedInputStream
import org.signal.core.util.stream.TrimmingInputStream
import org.signal.libsignal.protocol.InvalidMessageException
import org.signal.libsignal.protocol.incrementalmac.ChunkSizeChoice
import org.signal.libsignal.protocol.incrementalmac.IncrementalMacInputStream
import org.signal.libsignal.protocol.kdf.HKDF
import org.whispersystems.signalservice.api.backup.MediaRootBackupKey.MediaKeyMaterial
import org.whispersystems.signalservice.internal.util.Util
import java.io.ByteArrayInputStream
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.security.InvalidKeyException
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import javax.annotation.Nonnull
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlin.math.min

/**
 * Decrypts an attachment stream that has been encrypted with AES/CBC/PKCS5Padding.
 *
 * It assumes that the first 16 bytes of the stream are the IV, and that the rest of the stream is encrypted data.
 */
object AttachmentCipherInputStream {

  private const val BLOCK_SIZE = 16
  private const val CIPHER_KEY_SIZE = 32
  private const val MAC_KEY_SIZE = 32

  /**
   * Creates a stream to decrypt a typical attachment via a [File].
   *
   * @param incrementalDigest If null, incremental mac validation is disabled.
   * @param incrementalMacChunkSize If 0, incremental mac validation is disabled.
   */
  @JvmStatic
  @Throws(InvalidMessageException::class, IOException::class)
  fun createForAttachment(
    file: File,
    plaintextLength: Long,
    combinedKeyMaterial: ByteArray,
    integrityCheck: IntegrityCheck,
    incrementalDigest: ByteArray?,
    incrementalMacChunkSize: Int
  ): InputStream {
    return create(
      streamSupplier = { FileInputStream(file) },
      streamLength = file.length(),
      plaintextLength = plaintextLength,
      combinedKeyMaterial = combinedKeyMaterial,
      integrityCheck = integrityCheck,
      incrementalDigest = incrementalDigest,
      incrementalMacChunkSize = incrementalMacChunkSize
    )
  }

  /**
   * Creates a stream to decrypt a typical attachment via a [StreamSupplier].
   *
   * @param incrementalDigest If null, incremental mac validation is disabled.
   * @param incrementalMacChunkSize If 0, incremental mac validation is disabled.
   */
  @JvmStatic
  @Throws(InvalidMessageException::class, IOException::class)
  fun createForAttachment(
    streamSupplier: StreamSupplier,
    streamLength: Long,
    plaintextLength: Long,
    combinedKeyMaterial: ByteArray,
    integrityCheck: IntegrityCheck,
    incrementalDigest: ByteArray?,
    incrementalMacChunkSize: Int
  ): InputStream {
    return create(
      streamSupplier = streamSupplier,
      streamLength = streamLength,
      plaintextLength = plaintextLength,
      combinedKeyMaterial = combinedKeyMaterial,
      integrityCheck = integrityCheck,
      incrementalDigest = incrementalDigest,
      incrementalMacChunkSize = incrementalMacChunkSize
    )
  }

  /**
   * When you archive an attachment, you give the server an encrypted attachment, and the server wraps it in *another* layer of encryption.
   *
   * This creates a stream decrypt both the inner and outer layers of an archived attachment at the same time by basically double-decrypting it.
   *
   * @param incrementalDigest If null, incremental mac validation is disabled.
   * @param incrementalMacChunkSize If 0, incremental mac validation is disabled.
   */
  @JvmStatic
  @Throws(InvalidMessageException::class, IOException::class)
  fun createForArchivedMedia(
    archivedMediaKeyMaterial: MediaKeyMaterial,
    file: File,
    originalCipherTextLength: Long,
    plaintextLength: Long,
    combinedKeyMaterial: ByteArray,
    plaintextHash: ByteArray,
    incrementalDigest: ByteArray?,
    incrementalMacChunkSize: Int
  ): InputStream {
    val keyMaterial = CombinedKeyMaterial.from(combinedKeyMaterial)
    val mac = initMac(keyMaterial.macKey)

    if (originalCipherTextLength <= BLOCK_SIZE + mac.macLength) {
      throw InvalidMessageException("Message shorter than crypto overhead!")
    }

    return create(
      streamSupplier = { createForArchivedMediaOuterLayer(archivedMediaKeyMaterial, file, originalCipherTextLength) },
      streamLength = originalCipherTextLength,
      plaintextLength = plaintextLength,
      combinedKeyMaterial = combinedKeyMaterial,
      integrityCheck = IntegrityCheck(plaintextHash = plaintextHash, encryptedDigest = null),
      incrementalDigest = incrementalDigest,
      incrementalMacChunkSize = incrementalMacChunkSize
    )
  }

  /**
   * When you archive an attachment thumbnail, you give the server an encrypted attachment, and the server wraps it in *another* layer of encryption.
   *
   * This creates a stream decrypt both the inner and outer layers of an archived attachment at the same time by basically double-decrypting it.
   *
   * Archive thumbnails are also special in that we:
   * - don't know how long they are (meaning you'll get them back with padding at the end, image viewers are ok with this)
   * - don't care about external integrity checks (we still validate the MACs)
   *
   * So there's some code duplication here just to avoid mucking up the reusable functions with special cases.
   */
  @JvmStatic
  @Throws(InvalidMessageException::class, IOException::class)
  fun createForArchivedThumbnail(
    archivedMediaKeyMaterial: MediaKeyMaterial,
    file: File,
    innerCombinedKeyMaterial: ByteArray
  ): InputStream {
    val outerMac = initMac(archivedMediaKeyMaterial.macKey)

    if (file.length() <= BLOCK_SIZE + outerMac.macLength) {
      throw InvalidMessageException("Message shorter than crypto overhead! Expected at least ${BLOCK_SIZE + outerMac.macLength} bytes, got ${file.length()}")
    }

    FileInputStream(file).use { macVerificationStream ->
      verifyMacAndMaybeEncryptedDigest(macVerificationStream, file.length(), outerMac, null)
    }

    val outerEncryptedStreamExcludingMac = LimitedInputStream(FileInputStream(file), maxBytes = file.length() - outerMac.macLength)
    val outerCipher = createCipher(outerEncryptedStreamExcludingMac, archivedMediaKeyMaterial.aesKey)
    val innerEncryptedStream = BetterCipherInputStream(outerEncryptedStreamExcludingMac, outerCipher)

    val innerKeyMaterial = CombinedKeyMaterial.from(innerCombinedKeyMaterial)
    val innerMac = initMac(innerKeyMaterial.macKey)

    val innerEncryptedStreamWithMac = MacValidatingInputStream(innerEncryptedStream, innerMac)
    val innerEncryptedStreamExcludingMac = TrimmingInputStream(innerEncryptedStreamWithMac, trimSize = innerMac.macLength, drain = true)
    val innerCipher = createCipher(innerEncryptedStreamExcludingMac, innerKeyMaterial.aesKey)

    return BetterCipherInputStream(innerEncryptedStreamExcludingMac, innerCipher)
  }

  /**
   * Creates a stream to decrypt sticker data. Stickers have a special path because the key material is derived from the pack key.
   */
  @JvmStatic
  @Throws(InvalidMessageException::class, IOException::class)
  fun createForStickerData(data: ByteArray, packKey: ByteArray): InputStream {
    val keyMaterial = CombinedKeyMaterial.from(HKDF.deriveSecrets(packKey, "Sticker Pack".toByteArray(), 64))
    val mac = initMac(keyMaterial.macKey)

    if (data.size <= BLOCK_SIZE + mac.macLength) {
      throw InvalidMessageException("Message shorter than crypto overhead!")
    }

    ByteArrayInputStream(data).use { inputStream ->
      verifyMacAndMaybeEncryptedDigest(inputStream, data.size.toLong(), mac, null)
    }

    val encryptedStream = ByteArrayInputStream(data)
    val encryptedStreamExcludingMac = LimitedInputStream(encryptedStream, data.size.toLong() - mac.macLength)
    val cipher = createCipher(encryptedStreamExcludingMac, keyMaterial.aesKey)

    return BetterCipherInputStream(encryptedStreamExcludingMac, cipher)
  }

  /**
   * When you archive an attachment, you give the server an encrypted attachment, and the server wraps it in *another* layer of encryption.
   * This will return a stream that unwraps the server's layer of encryption, giving you a stream that contains a "normally-encrypted" attachment.
   *
   * Because we're validating the encryptedDigest/plaintextHash of the inner layer, there's no additional out-of-band validation of this outer layer.
   */
  @JvmStatic
  @Throws(InvalidMessageException::class, IOException::class)
  private fun createForArchivedMediaOuterLayer(archivedMediaKeyMaterial: MediaKeyMaterial, file: File, originalCipherTextLength: Long): LimitedInputStream {
    val mac = initMac(archivedMediaKeyMaterial.macKey)

    if (file.length() <= BLOCK_SIZE + mac.macLength) {
      throw InvalidMessageException("Message shorter than crypto overhead! Expected at least ${BLOCK_SIZE + mac.macLength} bytes, got ${file.length()}")
    }

    FileInputStream(file).use { macVerificationStream ->
      verifyMacAndMaybeEncryptedDigest(macVerificationStream, file.length(), mac, null)
    }

    val encryptedStream = FileInputStream(file)
    val encryptedStreamExcludingMac = LimitedInputStream(encryptedStream, file.length() - mac.macLength)
    val cipher = createCipher(encryptedStreamExcludingMac, archivedMediaKeyMaterial.aesKey)
    val inputStream: InputStream = BetterCipherInputStream(encryptedStreamExcludingMac, cipher)

    return LimitedInputStream(inputStream, originalCipherTextLength)
  }

  /**
   * @param integrityCheck If null, no integrity check is performed! This is a private method, so it's assumed that care has been taken to ensure that this is
   *   the correct course of action. Public methods should properly enforce when integrity checks are required.
   */
  @JvmStatic
  @Throws(InvalidMessageException::class, IOException::class)
  private fun create(
    streamSupplier: StreamSupplier,
    streamLength: Long,
    plaintextLength: Long,
    combinedKeyMaterial: ByteArray,
    integrityCheck: IntegrityCheck?,
    incrementalDigest: ByteArray?,
    incrementalMacChunkSize: Int
  ): InputStream {
    val keyMaterial = CombinedKeyMaterial.from(combinedKeyMaterial)
    val mac = initMac(keyMaterial.macKey)

    if (streamLength <= BLOCK_SIZE + mac.macLength) {
      throw InvalidMessageException("Message shorter than crypto overhead! length: $streamLength")
    }

    val wrappedStream: InputStream
    val hasIncrementalMac = incrementalDigest != null && incrementalDigest.isNotEmpty() && incrementalMacChunkSize > 0

    if (hasIncrementalMac) {
      if (integrityCheck == null) {
        throw InvalidMessageException("Missing integrityCheck for incremental mac validation!")
      }

      val digestValidatingStream = if (integrityCheck.encryptedDigest != null) {
        DigestValidatingInputStream(streamSupplier.openStream(), sha256Digest(), integrityCheck.encryptedDigest)
      } else {
        streamSupplier.openStream()
      }

      wrappedStream = IncrementalMacInputStream(
        IncrementalMacAdditionalValidationsInputStream(
          wrapped = digestValidatingStream,
          fileLength = streamLength,
          mac = mac
        ),
        keyMaterial.macKey,
        ChunkSizeChoice.everyNthByte(incrementalMacChunkSize),
        incrementalDigest
      )
    } else {
      streamSupplier.openStream().use { macVerificationStream ->
        verifyMacAndMaybeEncryptedDigest(macVerificationStream, streamLength, mac, integrityCheck?.encryptedDigest)
      }
      wrappedStream = streamSupplier.openStream()
    }

    val encryptedStreamExcludingMac = LimitedInputStream(wrappedStream, streamLength - mac.macLength)
    val cipher = createCipher(encryptedStreamExcludingMac, keyMaterial.aesKey)
    val decryptingStream: InputStream = BetterCipherInputStream(encryptedStreamExcludingMac, cipher)
    val paddinglessDecryptingStream = LimitedInputStream(decryptingStream, plaintextLength)

    return if (integrityCheck?.plaintextHash != null) {
      if (integrityCheck.plaintextHash.size != MessageDigest.getInstance("SHA-256").digestLength) {
        throw InvalidMessageException("Invalid plaintext hash size: ${integrityCheck.plaintextHash.size}")
      }

      DigestValidatingInputStream(paddinglessDecryptingStream, sha256Digest(), integrityCheck.plaintextHash)
    } else {
      paddinglessDecryptingStream
    }
  }

  private fun createCipher(inputStream: InputStream, aesKey: ByteArray): Cipher {
    val iv = inputStream.readNBytesOrThrow(BLOCK_SIZE)

    return Cipher.getInstance("AES/CBC/PKCS5Padding").apply {
      init(Cipher.DECRYPT_MODE, SecretKeySpec(aesKey, "AES"), IvParameterSpec(iv))
    }
  }

  private fun sha256Digest(): MessageDigest {
    try {
      return MessageDigest.getInstance("SHA-256")
    } catch (e: NoSuchAlgorithmException) {
      throw AssertionError(e)
    }
  }

  private fun initMac(key: ByteArray): Mac {
    try {
      val mac = Mac.getInstance("HmacSHA256")
      mac.init(SecretKeySpec(key, "HmacSHA256"))
      return mac
    } catch (e: NoSuchAlgorithmException) {
      throw AssertionError(e)
    } catch (e: InvalidKeyException) {
      throw AssertionError(e)
    }
  }

  @Throws(InvalidMessageException::class)
  private fun verifyMacAndMaybeEncryptedDigest(@Nonnull inputStream: InputStream, length: Long, @Nonnull mac: Mac, theirDigest: ByteArray?) {
    try {
      val digest = MessageDigest.getInstance("SHA256")
      var remainingData = Util.toIntExact(length) - mac.macLength
      val buffer = ByteArray(4096)

      while (remainingData > 0) {
        val read = inputStream.read(buffer, 0, min(buffer.size, remainingData))
        mac.update(buffer, 0, read)
        digest.update(buffer, 0, read)
        remainingData -= read
      }

      val ourMac = mac.doFinal()
      val theirMac = ByteArray(mac.macLength)
      Util.readFully(inputStream, theirMac)

      if (!MessageDigest.isEqual(ourMac, theirMac)) {
        throw InvalidMessageException("MAC doesn't match!")
      }

      val ourDigest = digest.digest(theirMac)

      if (theirDigest != null && !MessageDigest.isEqual(ourDigest, theirDigest)) {
        throw InvalidMessageException("Digest doesn't match!")
      }
    } catch (e: IOException) {
      throw InvalidMessageException(e)
    } catch (e: ArithmeticException) {
      throw InvalidMessageException(e)
    } catch (e: NoSuchAlgorithmException) {
      throw AssertionError(e)
    }
  }

  private class CombinedKeyMaterial(val aesKey: ByteArray, val macKey: ByteArray) {
    companion object {
      @Throws(InvalidMessageException::class)
      fun from(combinedKeyMaterial: ByteArray): CombinedKeyMaterial {
        if (combinedKeyMaterial.size != CIPHER_KEY_SIZE + MAC_KEY_SIZE) {
          throw InvalidMessageException("Invalid combined key material size: ${combinedKeyMaterial.size}")
        }
        val parts = Util.split(combinedKeyMaterial, CIPHER_KEY_SIZE, MAC_KEY_SIZE)
        return CombinedKeyMaterial(parts[0], parts[1])
      }
    }
  }

  fun interface StreamSupplier {
    @Nonnull
    @Throws(IOException::class)
    fun openStream(): InputStream
  }

  class IntegrityCheck(
    val encryptedDigest: ByteArray?,
    val plaintextHash: ByteArray?
  ) {
    init {
      if (encryptedDigest == null && plaintextHash == null) {
        throw IllegalArgumentException("At least one of encryptedDigest or plaintextHash must be provided")
      }
    }

    companion object {
      @JvmStatic
      fun forEncryptedDigest(encryptedDigest: ByteArray): IntegrityCheck {
        return IntegrityCheck(encryptedDigest, null)
      }

      @JvmStatic
      fun forPlaintextHash(plaintextHash: ByteArray): IntegrityCheck {
        return IntegrityCheck(null, plaintextHash)
      }

      @JvmStatic
      fun forEncryptedDigestAndPlaintextHash(encryptedDigest: ByteArray?, plaintextHash: String?): IntegrityCheck {
        val plaintextHashBytes = plaintextHash?.let { Base64.decode(it) }
        return IntegrityCheck(encryptedDigest, plaintextHashBytes)
      }
    }
  }
}
